File: Arrays/Rotate_An_Array/JS/RotateAnArray.js
const readline = require('readline');
const reverse = require('../Reverse_An_Array/ReverseAnArray');

function getUserInput(callback){

    reverse.getUserInput((arr) => {
        
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        rl.question('Enter the number of rotations ', (num_rotations) => {
            rl.question('Select Left / Right (L/R)?: ', (choice) => {
                rl.close();
                callback(arr, num_rotations, choice);
            });
        });
    });

}

function rotateAnArray(arr, a_end, b_start){
    // defining callback function to rotate the user entered array which is passed to this function as parameter
    reverse.reverseAnArray(arr, 0, a_end);
    reverse.reverseAnArray(arr, b_start, arr.length - 1)
    reverse.reverseAnArray(arr);
    reverse.printArray(arr);
}

 



 module.exports = {
    getUserInput: getUserInput,
    rotateAnArray: rotateAnArray
 } 

File: Arrays/Rotate_An_Array/Java/RotateAnArray.java
package Arrays.Rotate_An_Array.Java;

import java.util.Arrays;

import Arrays.Reverse_An_Array.Java.ReverseAnArray;
import Java_helper_classes.FastReader;

public class RotateAnArray extends ReverseAnArray {

    protected int num_rotations;
    protected String choice;

    public void getUserInputs() {

        getArrayFromUser();
        FastReader s = new FastReader();
        System.out.print("Enter the number of rotations you want to make: ");
        num_rotations = s.nextInt();
        System.out.print("Left/Right rotation (L/R) ?: ");
        choice = s.next();
    }


    public void rotateArray(){

        switch (choice.toUpperCase()) {
            case "L" :
                leftRotateAnArray(arr, num_rotations);
                break;
            case "R" :
                rightRotateAnArray(arr, num_rotations);
                break;
            default :
                System.out.println("You have entered an invalid input");
                return;
        }

    }

    public boolean isValidChoice(){
        return  choice.equalsIgnoreCase("L") || choice.equalsIgnoreCase("R");
    }

    private void leftRotateAnArray(int[] arr, int k) {
        // breaking arr into a(0 to k-1) and b(k to last) considering k as middle. reverse a and b then reverse entire to get the rotation
        reverseAnArray(0, k-1);
        reverseAnArray(k, arr.length-1);
        reverseAnArray(0, arr.length-1);
    }

    private void rightRotateAnArray(int[] arr, int k) {
        // breaking arr into a(0 to last-k) and b(k to last) considering k as middle. reverse a and b then reverse entire to get the rotation
        reverseAnArray(0, arr.length-k-1);
        reverseAnArray(arr.length-k, arr.length-1);
        reverseAnArray(0, arr.length-1);
    }

}
